src/installer/iohandler.py

```diff
@@ -0,0 +1,100 @@
+import os
+from io import RawIOBase
+
+from installer._compat.typing import TYPE_CHECKING
+
+from .records import Hash, Record
+
+if TYPE_CHECKING:
+    from ._compat.typing import FSPath
+    from typing import Dict, Tuple, Optional, Callable
+
+_BLOCK_SIZE = 4096
+_SCHEME_KEYS = ["platlib", "purelib"]
+
+
+def basic_copy_handler(
+    source,  # type: RawIOBase
+    dest,  # type: FSPath
+    hash_algorithm="sha256",  # type: Optional[str]
+):
+    # type: (...) -> Tuple[int, Hash]
+    """Copy a buffer and compute its hash and size.
+
+    Copies the source buffer to the specified destination. It also computes the hash
+    of the file to avoid traversing the source twice.
+
+    :param source: buffer holding the source data
+    :param dest: destination path
+    :param hash_algorithm: hashing algorithm
+
+    :return: size and ``Hash`` tuple of the copied buffer
+    """
+    assert hash_algorithm is not None  # to please mypy
+    hasher = hashlib.new(hash_algorithm)
+    size = 0
+    with open(dest, "wb") as dest_fd:
+        while True:
+            buf = source.read(_BLOCK_SIZE)
+            if not buf:
+                break
+            hasher.update(buf)
+            dest_fd.write(buf)
+            size += len(buf)
+
+    hash_ = Hash.parse("sha256={}".format(hasher.hexdigest()))
+    return size, hash_
+
+
+class IOHandler(object):
+    """Implement I/O operations for installers."""
+
+    def __init__(
+        self,
+        scheme_path_map,  # type: Dict[str, FSPath]
+        hash_algorithm="sha256",  # type: Optional[str]
+        copy_handler=basic_copy_handler,  # type: Callable[[RawIOBase, FSPath, Optional[str]], Tuple[int, Hash]]
+    ):
+        # type: (...) -> None
+        """Construct an ``IOHandler`` instance.
+
+        :param scheme_path_map: mapping of scheme names to paths
+        :param hash_algorithm: hashing algorithm passed to the ``copy_handler``
+        :param copy_handler: callable implementing the copy and hashing operation
+        """
+        self._scheme_path_map = scheme_path_map
+        self._hash_algo = hash_algorithm
+        self._copy_handler = copy_handler
+
+    def copy_file(
+        self,
+        scheme,  # type: str
+        path,  # type: FSPath
+        stream,  # type: RawIOBase
+    ):
+        # type: (...) -> Record
+        """Copy the provided path into the appropriate scheme location.
+
+        :param scheme: destination scheme
+        :param path: path inside the Python package
+        :param stream: source file contents
+
+        :return record of the installed file
+        """
+        try:
+            dest_dir = self._scheme_path_map[scheme]
+        except KeyError:
+            raise ValueError("No such scheme: {}".format(scheme))  # FIXME
+
+        dest_path = os.path.join(dest_dir, path)
+        parent_path = os.path.dirname(dest_path)
+
+        if not os.path.exists(parent_path):
+            # FIXME raise custom exception
+            os.makedirs(parent_path)
```

Comment:
I'd factor this out into a method, so that downstream could override it, and change behaviors?

Reply:
I wonder if we should just defer this to the copy handler. That way that implementation can decide how they want to deal with paths. For example, an (admittedly weird) implementation which wants to install wheels into a database could leave everything intact in `IOHandler` and only supply a different copy handler to achieve this.